File: android/app/src/test/java/com/nextpage/data/remote/catalog/CatalogSourceTest.kt
package com.nextpage.data.remote.catalog

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.fail
import org.junit.Test

/** Strict source-id parsing — mirrors desktop catalog-source.test.ts vectors. */
class CatalogSourceTest {

    private val addonId = "a1b2c3d4e5f60718"

    @Test
    fun `accepts the exact built-in ids`() {
        assertEquals(BUILTIN_GUTENDEX, parseCatalogSource("builtin:gutendex"))
        assertEquals(BUILTIN_OPENLIBRARY, parseCatalogSource("builtin:openlibrary"))
    }

    @Test
    fun `accepts addon sources with a 16-hex addonId`() {
        assertEquals("addon:$addonId", parseCatalogSource("addon:$addonId"))
        assertEquals("addon:$addonId", addonSource(addonId))
        assertEquals(addonId, addonSourceIdOf("addon:$addonId"))
    }

    @Test
    fun `addonSourceIdOf returns null for non-addon sources`() {
        assertNull(addonSourceIdOf(BUILTIN_GUTENDEX))
    }

    private fun expectRejected(raw: String) {
        try {
            parseCatalogSource(raw)
            fail("expected CatalogException for $raw")
        } catch (err: CatalogException) {
            assertEquals(CatalogErrorCode.NOT_FOUND, err.code)
        }
    }

    @Test
    fun `rejects bare names without a namespace prefix`() {
        expectRejected("gutendex")
        expectRejected("openlibrary")
    }

    @Test
    fun `rejects empty addon ids`() {
        expectRejected("addon:")
    }

    @Test
    fun `rejects unknown builtin ids`() {
        expectRejected("builtin:other")
    }

    @Test
    fun `rejects trailing junk and case drift on builtin ids`() {
        expectRejected("builtin:gutendex:x")
        expectRejected("builtin:gutendex ")
        expectRejected("builtin:GUTENDEX")
    }

    @Test
    fun `rejects malformed addon ids`() {
        expectRejected("addon:xyz")
        expectRejected("addon:ABCDEF0123456789")
        expectRejected("addon:${addonId}extra")
    }

    @Test
    fun `rejects unknown namespaces`() {
        expectRejected("curated:gutendex")
        expectRejected("")
    }
}

File: desktop/src/test/unit/services/catalog/catalog-source.test.ts
import { describe, expect, it } from 'vitest';
import { CatalogError } from '$lib/shared/services/catalog/errors';
import {
  BUILTIN_GUTENDEX,
  BUILTIN_OPENLIBRARY,
  addonSource,
  addonSourceIdOf,
  parseCatalogSource,
} from '$lib/shared/services/catalog/CatalogProvider';

const ADDON_ID = 'a1b2c3d4e5f60718';

describe('parseCatalogSource (strict source ids)', () => {
  it('accepts the exact built-in ids', () => {
    expect(parseCatalogSource('builtin:gutendex')).toBe(BUILTIN_GUTENDEX);
    expect(parseCatalogSource('builtin:openlibrary')).toBe(BUILTIN_OPENLIBRARY);
  });

  it('accepts addon sources with a 16-hex addonId', () => {
    expect(parseCatalogSource(`addon:${ADDON_ID}`)).toBe(`addon:${ADDON_ID}`);
    expect(addonSource(ADDON_ID)).toBe(`addon:${ADDON_ID}`);
    expect(addonSourceIdOf(`addon:${ADDON_ID}`)).toBe(ADDON_ID);
  });

  it('rejects bare names without a namespace prefix', () => {
    for (const raw of ['gutendex', 'openlibrary']) {
      expect(() => parseCatalogSource(raw)).toThrowError(CatalogError);
    }
  });

  it('rejects empty addon ids', () => {
    expect(() => parseCatalogSource('addon:')).toThrowError(CatalogError);
  });

  it('rejects unknown builtin ids (no tolerant parsing)', () => {
    expect(() => parseCatalogSource('builtin:other')).toThrowError(CatalogError);
  });

  it('rejects trailing junk on builtin ids', () => {
    expect(() => parseCatalogSource('builtin:gutendex:x')).toThrowError(CatalogError);
    expect(() => parseCatalogSource('builtin:gutendex ')).toThrowError(CatalogError);
    expect(() => parseCatalogSource('builtin:GUTENDEX')).toThrowError(CatalogError);
  });

  it('rejects malformed addon ids', () => {
    expect(() => parseCatalogSource('addon:xyz')).toThrowError(CatalogError);
    expect(() => parseCatalogSource('addon:ABCDEF0123456789')).toThrowError(CatalogError);
    expect(() => parseCatalogSource(`addon:${ADDON_ID}:extra`)).toThrowError(CatalogError);
  });

  it('rejects unknown namespaces', () => {
    expect(() => parseCatalogSource('curated:gutendex')).toThrowError(CatalogError);
    expect(() => parseCatalogSource('')).toThrowError(CatalogError);
  });
});
